### PR TITLE
[#1739] Add schema for json data files

### LINKED ISSUE #1739: Add schema for json data files
**What feature(s) would you like to see in RepoSense?**

I want to see json schema added for data files generated.

**Is the feature request related to a problem?**

It is troublesome troublesome to manually check what went wrong with the data file. Having a schema will allow easy debugging and see which specific field is missing/incorrect.

### PR DESCRIPTION
**Pull request by Zhou-Jiahao-1998**
Fixes #1739 

## Proposed commit message

```
There is no json schema for the data files generated. This can
be difficult to debug when the file is corrupted or when backend
generates a wrong json file.

Let us add schema file to our json files to make future debugging
easier.
```

## Other information

Refer to [this](https://www.jetbrains.com/help/idea/json.html#ws_json_using_schemas) for the guide on how to use the schema.

### CODE REVIEW

**Review by yhtMinceraft1010X — CHANGES_REQUESTED** [review]
This sounds like a great idea. I think the instructions on how to use this should be included in the DG, such as `workflow.md`.

**Comment by yhtMinceraft1010X** [comment]
One concern I have though is that the plugin described in the given guide is only available on IntelliJ IDEA Ultimate Edition, which is a paid product. Not every developer may have the paid version and consequently won't be able to make use of this feature.

**Comment by Zhou-Jiahao-1998** [comment]
> One concern I have though is that the plugin described in the given guide is only available on IntelliJ IDEA Ultimate Edition, which is a paid product. Not every developer may have the paid version and consequently won't be able to make use of this feature.

Yes that is a valid concern. I have added the instruction for VS Code in `workflow.md`.

**Review by yhtMinceraft1010X — APPROVED** [review]
LGTM

File: docs/dg/workflow.md [hunk 1]
```diff
@@ -89,6 +89,11 @@ The issues for first timers usually have guidance provided in the comment or hav
 See [vue-devtools project home](https://github.com/vuejs/vue-devtools) page for more details.
 </box>
 
+**You can use the json schemas provided in `/frontend/json-schema/` to check if the json data generated in `/reposense-report/` has the correct format.**
+* If you are using **IntelliJ**, please follow the guide [here](https://www.jetbrains.com/help/idea/json.html#ws_json_using_schemas).
+* If you are using **Visual Studio Code**, open the `.json` data file. Using `summary.json` as an example, add `"$schema": "../frontend/json-schema/summary-schema.json"` to the start of json. If you need to check using other schemas, replace `summary-schema.json` with other schema files provided.
```
Comment (yhtMinceraft1010X) [inline docs/dg/workflow.md:94]: Will adding the `$schema` line affect the way the frontend processes the json?

**Comment by gerhean** [comment]
Will this schema be used for typescript as well? I rather not have multiple schemas.

File: frontend/json-schema/authorship-schema.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{
```
Comment (gok99) [inline frontend/json-schema/authorship-schema.json:1]: I'm not sure if non-required fields need to be included here, but there is also a field `isBinary` (and `isIgnored` added in #1647) that are only serialized if true.

See: [FileResult.java](https://github.com/reposense/RepoSense/blob/2b864022bbf5db55ae9d4f96eaf0a70e961646f5/src/main/java/reposense/authorship/model/FileResult.java)

**Review by gok99 — COMMENTED** [review]
Just had a thought (this could probably be explored separately if at all): would it be a good idea to generate these automatically from the Java classes which can act as a single truth source? Haven't looked into this too much but maybe this is useful: http://wilddiary.com/generate-json-schema-from-java-class/

**Comment by gerhean** [comment]
I agree with @gok99 that I would rather have a single file which both the frontend and java can verify json against.

However my priority for now is Typescript integration.

**Pull request closed without merging**